protocol/serialization-types-macros/src/auto_from.rs

```diff
@@ -0,0 +1,368 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+//!
+//! Utilities for implementing AutoFrom
+//!
+
+use crate::*;
+use syn::{punctuated::Punctuated, AttrStyle, Attribute, Field, Fields, Token, Variant};
+
+/// Parse the attributed target type(s)
+pub fn parse_types_from_attr(attributes: &[Attribute]) -> Vec<proc_macro2::TokenStream> {
+    let mut target_types: Vec<proc_macro2::TokenStream> = Vec::new();
+    for attr in attributes {
+        if let AttrStyle::Outer = attr.style {
+            if let Some(attr_id) = attr.path.get_ident() {
+                if &attr_id.to_string() == "auto_from" {
+                    if let Ok(args) = attr.parse_args::<proc_macro2::TokenStream>() {
+                        target_types.push(args.clone());
+                    }
+                }
+            }
+        };
+    }
+    target_types
+}
+
+/// Codegen for structs with named fields
+pub fn auto_from_for_struct_with_named_fields(
+    ident: &proc_macro2::Ident,
+    target_types: &[proc_macro2::TokenStream],
+    named: Punctuated<Field, Token![,]>,
+) -> Option<TokenStream> {
+    let mut field_idents = Vec::new();
+    let mut vec_field_idents = Vec::new();
+    let mut option_field_idents = Vec::new();
+    let mut boxed_field_idents = Vec::new();
+    'outer: for f in named {
+        if let Some(ident) = f.ident {
+            if let syn::Type::Path(type_path) = f.ty {
+                for seg in type_path.path.segments {
+                    match seg.ident.to_string().as_str() {
+                        "Vec" => {
+                            vec_field_idents.push(ident);
+                            continue 'outer;
+                        }
+                        "Option" => {
+                            option_field_idents.push(ident);
+                            continue 'outer;
+                        }
+                        "Box" => {
+                            boxed_field_idents.push(ident);
+                            continue 'outer;
+                        }
+                        _ => {}
```

Comment:
Would be good to log the ident here, for debugging purposes

Reply:
This default variant here is the most common case that falls to the default handler, adding `println!` here will flood build output